addons/website/models/website_visitor.py

```diff
@@ -0,0 +1,177 @@
+            return "%s sec ago" % round(seconds)
+        elif seconds / (60 * 60) <= 1:
+            return "%s min ago" % round(seconds / 60)
+        elif seconds / (60 * 60 * 24) <= 1:
+            return "%s h ago" % round(seconds / (60*60))
+        else:
+            nbr_days = round(seconds / (60 * 60 * 24))
+            return "%s day%s ago" % (nbr_days, '' if nbr_days == 1 else 's')
+
+    def _encode(self, visitor_id):
+        return "%d-%s" % (visitor_id, self._get_visitor_hash(visitor_id))
+
+    def _get_visitor_hash(self, visitor_id):
+        return md5(("%s%s" % (visitor_id, self.env['ir.config_parameter'].sudo().get_param('database.secret'))).encode('utf-8')).hexdigest()
+
+    def _get_visitor_from_request(self):
+        if not request:
+            return None
+        visitor = self.env['website.visitor']
+        cookie_content = request.httprequest.cookies.get('visitor_id')
+        if cookie_content and '-' in cookie_content:
+            visitor_id, visitor_hash = cookie_content.split('-', 1)
+            if _consteq(visitor_hash, self._get_visitor_hash(visitor_id)):
+                return visitor.sudo().with_context(active_test=False).search([('id', '=', visitor_id)])  # search to avoid having to call exists()
+        return visitor
+
+    def _handle_webpage_dispatch(self, response, is_webpage=False):
+        """ Called on dispatch. This will create a website.visitor if the http request object
+        is a tracked website page. Only on tracked page to avoid having too much operations done on every page
+        or other http requests.
+        Note: The side effect is that the last_connection_datetime is updated ONLY on tracked pages."""
+        website_page = is_webpage and response.qcontext['main_object']
+        if website_page and website_page.is_tracked:
+            visitor_sudo = self._get_visitor_from_request()
+            if not visitor_sudo:
+                # If visitor does not exist
+                visitor_sudo = self._create_visitor(website_page.id)
+                sign = self._encode(visitor_sudo.id)
+                response.set_cookie('visitor_id', sign)
+            else:
+                # Add page even if already in visitor_page_ids as checks on relations are done in many2many write method
+                vals = {
+                    'last_connection_datetime': datetime.now(),
+                    'visitor_page_ids': [(0, 0, {'page_id': website_page.id, 'visit_date': datetime.now()})],
+                }
+                if visitor_sudo.last_connection_datetime < (datetime.now() - timedelta(hours=8)):
+                    vals['visit_count'] = visitor_sudo.visit_count + 1
+                if not visitor_sudo.active:
+                    vals['active'] = True
+                visitor_sudo.write(vals)
+        return response
+
+    def _create_visitor(self, website_page_id=False):
+        country_code = request.session.geoip and request.session.geoip.get('country_code')
```

Comment:
Will there always be a `.get('country_code')` and a `request.lang` ?
Shouldn't you check that they actually return a value before searching the ids?